At the time of scram build it reports:

>> Building shared library tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/libGeometryCaloTopology.so
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/slc7_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-unknown-linux-gnu/11.2.1/../../../../x86_64-unknown-linux-gnu/bin/ld: tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/HGCalTBTopology.cc.o: in function `std::type_info const& edm::typelookup::classTypeInfo<HGCalTBTopology>()':
HGCalTBTopology.cc:(.text+0x1f3): undefined reference to `typeinfo for HGCalTBTopology'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/slc7_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-unknown-linux-gnu/11.2.1/../../../../x86_64-unknown-linux-gnu/bin/ld: tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/HGCalTBTopology.cc.o: in function `HGCalTBTopology::HGCalTBTopology(HGCalTBDDDConstants const&, int)':
HGCalTBTopology.cc:(.text+0x467): undefined reference to `vtable for HGCalTBTopology'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/slc7_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-unknown-linux-gnu/11.2.1/../../../../x86_64-unknown-linux-gnu/bin/ld: tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/HGCalTBTopology.cc.o: in function `_GLOBAL__sub_I_HGCalTBTopology.cc':
HGCalTBTopology.cc:(.text.startup+0xee): undefined reference to `typeinfo for HGCalTBTopology'
collect2: error: ld returned 1 exit status
gmake: *** [config/SCRAM/GMake/Makefile.rules:1717: tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/libGeometryCaloTopology.so] Error 1
gmake: *** Waiting for unfinished jobs....
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2


________________________________
From: cmsbuild ***@***.***
Sent: 24 January 2023 18:02
To: cms-sw/cmssw
Cc: Sunanda Banerjee; Mention
Subject: Re: [cms-sw/cmssw] Phase2-hgx335J Try to add toplogy class for HGCal TB application (PR #40603)


A new Pull Request was created by @bsunanda<https://github.com/bsunanda> (Sunanda Banerjee) for master.

It involves the following packages:

  *   Geometry/CaloTopology (geometry)

@civanch<https://github.com/civanch>, @Dr15Jones<https://github.com/Dr15Jones>, @makortel<https://github.com/makortel>, @mdhildreth<https://github.com/mdhildreth>, @cmsbuild<https://github.com/cmsbuild>, @bsunanda<https://github.com/bsunanda> can you please review it and eventually sign? Thanks.
@simonepigazzini<https://github.com/simonepigazzini>, @fabiocos<https://github.com/fabiocos>, @thomreis<https://github.com/thomreis> this is something you requested to watch as well.
@perrotta<https://github.com/perrotta>, @dpiparo<https://github.com/dpiparo>, @rappoccio<https://github.com/rappoccio> you are the release manager for this.

cms-bot commands are listed here<http://cms-sw.github.io/cms-bot-cmssw-cmds.html>

—
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/40603#issuecomment-1402281657>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABGMZORGQMWGTNWG5OLBSATWUADL3ANCNFSM6AAAAAAUFKBBLI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

--- a/Geometry/CaloEventSetup/plugins/HGCalTBTopologyBuilder.cc
+++ b/Geometry/CaloEventSetup/plugins/HGCalTBTopologyBuilder.cc
@@ -76,7 +76,7 @@ HGCalTBTopologyBuilder::ReturnType HGCalTBTopologyBuilder::produce(const IdealGe
 #ifdef EDM_ML_DEBUG
   edm::LogVerbatim("HGCalGeom") << "Create HGCalTBTopology(hgdc,det)";
 #endif
-  return std::make_unique<HGCalTBTopology>(hgdc, det_);
+  return std::make_unique<HGCalTBTopology>(&hgdc, det_);
 }
 
 DEFINE_FWK_EVENTSETUP_MODULE(HGCalTBTopologyBuilder);

--- a/Geometry/CaloEventSetup/plugins/HGCalTBTopologyBuilder.cc
+++ b/Geometry/CaloEventSetup/plugins/HGCalTBTopologyBuilder.cc
@@ -1,0 +1,82 @@
+// -*- C++ -*-
+//
+// Package:    CaloEventSetup
+// Class:      HGCalTBTopologyBuilder
+//
+/**\class HGCalTBTopologyBuilder HGCalTBTopologyBuilder.h 
+
+ Description: <one line class summary>
+
+ Implementation:
+     <Notes on implementation>
+*/
+//
+// Original Author:  Sunanda Banerjee
+//
+//
+
+// system include files
+#include <memory>
+
+// user include files
+#include <FWCore/Framework/interface/ModuleFactory.h>
+#include "FWCore/Framework/interface/ESProducer.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/ESGetToken.h"
+
+#include "Geometry/CaloTopology/interface/HGCalTBTopology.h"
+#include "Geometry/CaloTopology/interface/CaloSubdetectorTopology.h"
+#include "Geometry/HGCalTBCommonData/interface/HGCalTBDDDConstants.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/CaloTopology/interface/CaloSubdetectorTopology.h"
+#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+
+//#define EDM_ML_DEBUG
+
+//
+// class decleration
+//
+
+class HGCalTBTopologyBuilder : public edm::ESProducer {
+public:
+  HGCalTBTopologyBuilder(const edm::ParameterSet& iP);
+  ~HGCalTBTopologyBuilder() override;
+
+  using ReturnType = std::unique_ptr<HGCalTBTopology>;
+
+  ReturnType produce(const IdealGeometryRecord&);
+
+private:
+  // ----------member data ---------------------------
+  edm::ESGetToken<HGCalTBDDDConstants, IdealGeometryRecord> hgcToken_;
+  int det_;
+};
+
+HGCalTBTopologyBuilder::HGCalTBTopologyBuilder(const edm::ParameterSet& iConfig) {
+  auto name = iConfig.getParameter<std::string>("Name");
+  det_ = iConfig.getParameter<int>("Type");
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCalGeom") << "constructing HGCalTBTopology for " << name << " and det " << det_;
+#endif
+  hgcToken_ = setWhatProduced(this, name).consumes<HGCalTBDDDConstants>(edm::ESInputTag{"", name});
+}
+
+HGCalTBTopologyBuilder::~HGCalTBTopologyBuilder() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+HGCalTBTopologyBuilder::ReturnType HGCalTBTopologyBuilder::produce(const IdealGeometryRecord& iRecord) {
+  const HGCalTBDDDConstants& hgdc = iRecord.get(hgcToken_);
+
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCalGeom") << "Create HGCalTBTopology(hgdc,det)";
+#endif
+  return std::make_unique<HGCalTBTopology>(hgdc, det_);
+}
+
+DEFINE_FWK_EVENTSETUP_MODULE(HGCalTBTopologyBuilder);

--- a/Geometry/CaloEventSetup/python/HGCalTBTopology_cfi.py
+++ b/Geometry/CaloEventSetup/python/HGCalTBTopology_cfi.py
@@ -1,0 +1,14 @@
+import FWCore.ParameterSet.Config as cms
+
+#
+# This cfi should be included to build the HGCal TB Topologies
+#
+HGCalTBEETopologyBuilder = cms.ESProducer("HGCalTBTopologyBuilder",
+                                          Name = cms.string("HGCalEESensitive"),
+                                          Type = cms.int32(3) )
+
+
+HGCalTBHESilTopologyBuilder = cms.ESProducer("HGCalTBTopologyBuilder",
+                                             Name = cms.string("HGCalHESiliconSensitive"),
+                                             Type = cms.int32(4) )
+

--- a/Geometry/CaloTopology/interface/HGCalTBTopology.h
+++ b/Geometry/CaloTopology/interface/HGCalTBTopology.h
@@ -12,10 +12,10 @@
 class HGCalTBTopology : public CaloSubdetectorTopology {
 public:
   /// create a new Topology
+  HGCalTBTopology() {}
   HGCalTBTopology(const HGCalTBDDDConstants& hdcons, int subdet);
 
-  /// virtual destructor
-  ~HGCalTBTopology() override = default;
+  /// default destructor
 
   /// move the Topology north (increment iy)
   DetId goNorth(const DetId& id) const override { return changeXY(id, 0, +1); }

--- a/Geometry/CaloTopology/interface/HGCalTBTopology.h
+++ b/Geometry/CaloTopology/interface/HGCalTBTopology.h
@@ -12,8 +12,7 @@
 class HGCalTBTopology : public CaloSubdetectorTopology {
 public:
   /// create a new Topology
-  HGCalTBTopology() {}
-  HGCalTBTopology(const HGCalTBDDDConstants& hdcons, int subdet);
+  HGCalTBTopology(const HGCalTBDDDConstants* hdcons, int subdet);
 
   /// default destructor
 
@@ -89,7 +88,7 @@ public:
   unsigned int totalGeomModules() const { return (unsigned int)(2 * kHGeomHalf_); }
   unsigned int allGeomModules() const;
 
-  const HGCalTBDDDConstants& dddConstants() const { return hdcons_; }
+  const HGCalTBDDDConstants& dddConstants() const { return *hdcons_; }
 
   /** returns a new DetId offset by nrStepsX and nrStepsY (can be negative),
    * returns DetId(0) if invalid */
@@ -113,13 +112,15 @@ public:
   bool detectorType() const { return false; }
 
 private:
+  HGCalTBTopology() : hdcons_(nullptr) {}
+
   /// move the nagivator along x, y
   DetId changeXY(const DetId& id, int nrStepsX, int nrStepsY) const;
 
   /// move the nagivator along z
   DetId changeZ(const DetId& id, int nrStepsZ) const;
 
-  const HGCalTBDDDConstants& hdcons_;
+  const HGCalTBDDDConstants* hdcons_;
   HGCalGeometryMode::GeometryMode mode_;
 
   DetId::Detector det_;

--- a/Geometry/CaloTopology/interface/HGCalTBTopology.h
+++ b/Geometry/CaloTopology/interface/HGCalTBTopology.h
@@ -112,8 +112,6 @@ public:
   bool detectorType() const { return false; }
 
 private:
-  HGCalTBTopology() : hdcons_(nullptr) {}
-
   /// move the nagivator along x, y
   DetId changeXY(const DetId& id, int nrStepsX, int nrStepsY) const;
 

--- a/Geometry/CaloTopology/src/HGCalTBTopology.cc
+++ b/Geometry/CaloTopology/src/HGCalTBTopology.cc
@@ -13,7 +13,7 @@ HGCalTBTopology::HGCalTBTopology(const HGCalTBDDDConstants* hdcons, int det) : h
   kHGhalf_ = sectors_ * layers_ * cells_;
   firstLay_ = hdcons_->firstLayer();
   det_ = DetId::Forward;
-  subdet_ = (ForwardSubdetector)(det);
+  subdet_ = static_cast<ForwardSubdetector>(det);
   kHGeomHalf_ = sectors_ * layers_;
   types_ = 2;
   kHGhalfType_ = sectors_ * layers_ * cells_ * types_;

--- a/Geometry/CaloTopology/src/HGCalTBTopology.cc
+++ b/Geometry/CaloTopology/src/HGCalTBTopology.cc
@@ -5,13 +5,13 @@
 
 #define EDM_ML_DEBUG
 
-HGCalTBTopology::HGCalTBTopology(const HGCalTBDDDConstants& hdcons, int det) : hdcons_(hdcons) {
-  sectors_ = hdcons_.sectors();
-  layers_ = hdcons_.layers(true);
-  cells_ = hdcons_.maxCells(true);
-  mode_ = hdcons_.geomMode();
+HGCalTBTopology::HGCalTBTopology(const HGCalTBDDDConstants* hdcons, int det) : hdcons_(hdcons) {
+  sectors_ = hdcons_->sectors();
+  layers_ = hdcons_->layers(true);
+  cells_ = hdcons_->maxCells(true);
+  mode_ = hdcons_->geomMode();
   kHGhalf_ = sectors_ * layers_ * cells_;
-  firstLay_ = hdcons_.firstLayer();
+  firstLay_ = hdcons_->firstLayer();
   det_ = DetId::Forward;
   subdet_ = (ForwardSubdetector)(det);
   kHGeomHalf_ = sectors_ * layers_;
@@ -26,7 +26,7 @@ HGCalTBTopology::HGCalTBTopology(const HGCalTBDDDConstants& hdcons, int det) : h
 #endif
 }
 
-unsigned int HGCalTBTopology::allGeomModules() const { return (static_cast<unsigned int>(2 * hdcons_.wafers())); }
+unsigned int HGCalTBTopology::allGeomModules() const { return (static_cast<unsigned int>(2 * hdcons_->wafers())); }
 
 DetId HGCalTBTopology::denseId2detId(uint32_t hi) const {
   HGCalTBTopology::DecodedDetId id;
@@ -63,7 +63,7 @@ bool HGCalTBTopology::valid(const DetId& idin) const {
   flag = (idin.det() == det_ && idin.subdetId() == (int)(subdet_) && id.iCell1 >= 0 && id.iCell1 < cells_ &&
           id.iLay > 0 && id.iLay <= layers_ && id.iSec1 >= 0 && id.iSec1 <= sectors_);
   if (flag)
-    flag = hdcons_.isValidHex(id.iLay, id.iSec1, id.iCell1, true);
+    flag = hdcons_->isValidHex(id.iLay, id.iSec1, id.iCell1, true);
   return flag;
 }
 

